src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs

```diff
@@ -234,7 +234,27 @@ internal QueryNode BindFunctionCall(FunctionCallToken functionCallToken)
 
             // If there isn't, bind as Uri function
             // Bind all arguments
-            List<QueryNode> argumentNodes = new List<QueryNode>(functionCallToken.Arguments.Select(ar => this.bindMethod(ar)));
+            List<QueryNode> argumentNodes = functionCallToken.Arguments.Select(argument =>
+            {
+                // If the function is IsOf or Cast and the argument is a dotted identifier, we need to bind it differently
+                if (UnboundFunctionNames.Contains(functionCallToken.Name) && argument.ValueToken is DottedIdentifierToken dottedIdentifier)
+                {
+                    // Find the type of the dotted identifier by resolving it against the model. This also ensure case-insensitive resolution.
+                    IEdmSchemaType dottedIdentifierType = UriEdmHelpers.FindTypeFromModel(state.Model, dottedIdentifier.Identifier, this.Resolver);
+
+                    // If the dotted identifier is a primitive type and the next token is not null, we need to ensure next token is null
+                    if (dottedIdentifierType is IEdmPrimitiveType && dottedIdentifier.NextToken != null)

```

Comment:
Why do we need to ensure that the next token is null?

Reply:
Consider the following:
- `cast` or `isof` with 2 parameters. For example, cast(ID, Edm.String):
   - The parser will create a `FunctionCallToken` with 2 arguments where the first argument (`ID`) is an `EndPathToken` and the second argument (`Edm.String`) is a `DottedIdentifierToken` because it is unquoted. When the type parameter is quoted, the second argument becomes a `LiteralToken`.
   - The second argument's Next Token will point to the first argument. This is important for determining if the type of the second argument can be assigned from the type of the first argument.
   - For primitive DottedIdentifier types like Edm.String, Edm.Int32, Edm.Byte (including their lowercase variants), the presence of NextToken will cause the DottedIdentifier to bind as a `SingleValueCastNode`, leading to potential breaking changes as reported in issue https://github.com/OData/odata.net/issues/3123
   
- `cast` or `isof` with 1 parameter. For example, cast(Edm.String):
  - The parser will create a `FunctionCallToken` with 1 argument where the argument (`Edm.String`) is a `DottedIdentifierToken`. This argument does not have a NextToken.

Here we are checking DottedIdentifier is primitive type and next token is null to ensure we handle the primitive types properly and cases where argument is 1.